From c3aacf332fccdb78085cb32d44d3e984060e335a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kacper=20Michaj=C5=82ow?= <kasper93@gmail.com>
Date: Thu, 14 May 2026 20:58:02 +0200
Subject: [PATCH] colorspace: do not force PQ transfer to zero nits black
 always

This is not the correct place to make this adjustment. It causes Linear
-> PQ tone-mapping to always trigger, even when the HDR metadata is
otherwise unchanged, which should be a no-op.

In libplacebo, not only PQ is scaled to absolute black. Linear and scRGB
are also affected, both of which correctly respect HDR metadata.

This change makes it so that PQ transfer is no longer hardcoded to
target infinite-contrast (virtual) displays.

This change will cause user metadata to be used, which may result in
elevated black levels in PQ output if it is not set correctly. The
mitigation is, however, easy. Setting `hdr.min_luma =
PL_COLOR_HDR_BLACK` restores the previous behavior. However, it is
recommended to use a value better matching your display.

Note that even OLEDs are not infinite-contrast in practice, because
while individual pixels may be fully disabled, panel reflectance can
still raise the black point, especially on QD-OLED panels.
---
 src/colorspace.c       |  4 ----
 src/tests/colorspace.c | 16 ++++++++--------
 2 files changed, 8 insertions(+), 12 deletions(-)

diff --git a/src/colorspace.c b/src/colorspace.c
index cfc9f401..29e6b5a9 100644
--- a/src/colorspace.c
+++ b/src/colorspace.c
@@ -823,10 +823,6 @@ void pl_color_space_nominal_luma_ex(const struct pl_nominal_luma_params *params)
     if ((max_luma && min_luma >= max_luma) || min_luma >= hdr_max)
         min_luma = max_luma = 0; // sanity
 
-    // PQ is always scaled down to absolute black, ignoring HDR metadata
-    if (csp->transfer == PL_COLOR_TRC_PQ)
-        min_luma = hdr_min;
-
     // Baseline/fallback metadata, inferred entirely from the colorspace
     // description and built-in default assumptions
     if (!max_luma) {
diff --git a/src/tests/colorspace.c b/src/tests/colorspace.c
index c13b456d..7112ffba 100644
--- a/src/tests/colorspace.c
+++ b/src/tests/colorspace.c
@@ -496,10 +496,10 @@ int main()
     REQUIRE(!pl_hdr_metadata_contains(&hdr10plus.hdr, PL_HDR_METADATA_CIE_Y));
 
     TEST_METADATA(hdr10plus, PL_HDR_METADATA_NONE,      PL_COLOR_HDR_BLACK, 10000, 0);
-    TEST_METADATA(hdr10plus, PL_HDR_METADATA_CIE_Y,     PL_COLOR_HDR_BLACK, 4000, 0);
-    TEST_METADATA(hdr10plus, PL_HDR_METADATA_HDR10,     PL_COLOR_HDR_BLACK, 4000, 0);
-    TEST_METADATA(hdr10plus, PL_HDR_METADATA_HDR10PLUS, PL_COLOR_HDR_BLACK, 1000, 250);
-    TEST_METADATA(hdr10plus, PL_HDR_METADATA_ANY,       PL_COLOR_HDR_BLACK, 1000, 250);
+    TEST_METADATA(hdr10plus, PL_HDR_METADATA_CIE_Y,     0.005, 4000, 0);
+    TEST_METADATA(hdr10plus, PL_HDR_METADATA_HDR10,     0.005, 4000, 0);
+    TEST_METADATA(hdr10plus, PL_HDR_METADATA_HDR10PLUS, 0.005, 1000, 250);
+    TEST_METADATA(hdr10plus, PL_HDR_METADATA_ANY,       0.005, 1000, 250);
 
     const struct pl_color_space dovi = {
         .primaries = PL_COLOR_PRIM_BT_2020,
@@ -519,10 +519,10 @@ int main()
     REQUIRE(!pl_hdr_metadata_contains(&dovi.hdr, PL_HDR_METADATA_HDR10PLUS));
 
     TEST_METADATA(dovi, PL_HDR_METADATA_NONE,      PL_COLOR_HDR_BLACK, 10000, 0);
-    TEST_METADATA(dovi, PL_HDR_METADATA_HDR10,     PL_COLOR_HDR_BLACK, 4000, 0);
-    TEST_METADATA(dovi, PL_HDR_METADATA_HDR10PLUS, PL_COLOR_HDR_BLACK, 4000, 0);
-    TEST_METADATA(dovi, PL_HDR_METADATA_CIE_Y,     PL_COLOR_HDR_BLACK, 1000, 250);
-    TEST_METADATA(dovi, PL_HDR_METADATA_ANY,       PL_COLOR_HDR_BLACK, 1000, 250);
+    TEST_METADATA(dovi, PL_HDR_METADATA_HDR10,     0.005, 4000, 0);
+    TEST_METADATA(dovi, PL_HDR_METADATA_HDR10PLUS, 0.005, 4000, 0);
+    TEST_METADATA(dovi, PL_HDR_METADATA_CIE_Y,     0.005, 1000, 250);
+    TEST_METADATA(dovi, PL_HDR_METADATA_ANY,       0.005, 1000, 250);
 
     const struct pl_color_space hlg4000 = {
         .primaries    = PL_COLOR_PRIM_BT_2020,
-- 
GitLab

